This is inspired from the Debian GHC patch "no-pie", found in:
http://deb.debian.org/debian/pool/main/g/ghc/ghc_8.0.1-17.debian.tar.xz
More information there, but this basically fixes linking of some haskell
components on recent compilers/distros.

--- a/compiler/main/DynFlags.hs
+++ b/compiler/main/DynFlags.hs
@@ -2332,7 +2332,8 @@ machdepCCOpts _dflags
                         -- built-in functions like memcpy() it tends to
                         -- run out of registers, requiring -monly-n-regs
                         "-fno-builtin",
-                        "-DSTOLEN_X86_REGS="++show n_regs ]
+                        "-DSTOLEN_X86_REGS="++show n_regs,
+                        "-no-pie" ]
                     )
 
 #elif ia64_TARGET_ARCH
@@ -2351,10 +2352,11 @@ machdepCCOpts _dflags
                         -- and get in the way of -split-objs.  Another option
                         -- would be to throw them away in the mangler, but this
                         -- is easier.
-                 "-fno-builtin"
+                 "-fno-builtin",
                         -- calling builtins like strlen() using the FFI can
                         -- cause gcc to run out of regs, so use the external
                         -- version.
+                 "-no-pie"
                 ] )
 
 #elif sparc_TARGET_ARCH
--- a/compiler/main/DriverPipeline.hs
+++ b/compiler/main/DriverPipeline.hs
@@ -1685,7 +1685,7 @@ linkDynLib dflags o_files dep_packages =
 	 ++ map SysTools.Option (
 	    md_c_flags
 	 ++ o_files
-	 ++ [ "-shared", "-Wl,-Bsymbolic" ] -- we need symbolic linking to resolve non-PIC intra-package-relocations
+	 ++ [ "-no-pie", "-shared", "-Wl,-Bsymbolic" ] -- we need symbolic linking to resolve non-PIC intra-package-relocations
          ++ [ "-Wl,-soname," ++ takeFileName output_fn ] -- set the library soname
 	 ++ extra_ld_inputs
 	 ++ lib_path_opts
